File: app/src/main/java/com/afoxxvi/alopex/component/filter/AlopexFilterManager.java
package com.afoxxvi.alopex.component.filter;

import static com.afoxxvi.alopex.AlopexView.TAG;

import android.content.Context;
import android.util.Log;

import com.afoxxvi.alopex.component.notify.Notify;
import com.afoxxvi.alopex.util.FileUtils;
import com.afoxxvi.alopex.util.Triplet;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class AlopexFilterManager {
    private AlopexFilterManager() {
    }

    private static AlopexFilterManager inst;

    public static AlopexFilterManager getInstance() {
        if (inst == null) {
            inst = new AlopexFilterManager();
        }
        return inst;
    }

    public void init(Context context) {
        filters = new ArrayList<>();
        File file = new File(context.getFilesDir(), "filter.json");
        if (!file.exists()) {
            try {
                file.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            String doc = FileUtils.inputFile(context, "filter.json");
            try {
                JSONArray array = new JSONArray(doc);
                for (int i = 0; i < array.length(); i++) {
                    JSONObject object = array.optJSONObject(i);
                    if (object != null) {
                        filters.add(AlopexFilter.fromJsonObject(object));
                    }
                }
            } catch (JSONException ignored) {
            }
        }
    }

    public void save(Context context) {
        JSONArray array = new JSONArray();
        for (AlopexFilter filter : filters) {
            array.put(filter.toJsonObject());
        }
        FileUtils.outputFile(context, "filter.json", array.toString());
        Log.i(TAG, "filter saved");
    }

    private List<AlopexFilter> filters;

    public List<AlopexFilter> getFilters() {
        return filters;
    }


    public Triplet<Boolean, Boolean, Boolean> isFiltered(String packageName, Notify notify, boolean count) {
        return isFiltered(packageName, notify.getTitle(), notify.getText(), count);
    }

    /**
     * @return do notify, do cancel, is filtered
     */
    public Triplet<Boolean, Boolean, Boolean> isFiltered(String packageName, String title, String text, boolean count) {
        for (AlopexFilter filter : filters) {
            if (filter.getPackageName().equals(packageName)) {
                if (filter.isFiltered(title, text, count)) {
                    return new Triplet<>(false, filter.getCancelFiltered(), true);
                } else {
                    return new Triplet<>(filter.getNotifyUnfiltered(), false, false);
                }
            }
        }
        return new Triplet<>(false, false, false);
    }
}

File: app/src/main/java/com/afoxxvi/alopex/component/filter/AlopexFilterManager.kt
package com.afoxxvi.alopex.component.filter

import android.content.Context
import android.util.Log
import com.afoxxvi.alopex.Alopex
import com.afoxxvi.alopex.component.filter.AlopexFilter.Companion.fromJsonObject
import com.afoxxvi.alopex.component.notify.Notify
import com.afoxxvi.alopex.util.FoxFiles.inputFile
import com.afoxxvi.alopex.util.FoxFiles.outputFile
import com.afoxxvi.alopex.util.FoxTools.singleExecutorService
import com.afoxxvi.alopex.util.Triplet
import org.json.JSONArray
import java.io.File
import java.util.concurrent.TimeUnit

object AlopexFilterManager {
    private var filterList: MutableList<AlopexFilter> = mutableListOf()
    val filters: MutableList<AlopexFilter>
        get() {
            return filterList
        }

    fun init(context: Context) {
        filters.clear()
        val file = File(context.filesDir, "filter.json")
        if (!file.exists()) {
            file.createNewFile()
        } else {
            val doc = inputFile(context, "filter.json")
            val array = JSONArray(doc)
            for (i in 0 until array.length()) {
                val obj = array.optJSONObject(i)
                if (obj != null) {
                    filters.add(fromJsonObject(obj))
                }
            }
        }
        singleExecutorService.scheduleWithFixedDelay({ save(context) }, 3, 300, TimeUnit.SECONDS)
    }

    fun save(context: Context?) {
        val array = JSONArray()
        for (filter in filters) {
            array.put(filter.toJsonObject())
        }
        outputFile(context!!, "filter.json", array.toString())
        Log.i(Alopex.TAG, "filter saved")
    }

    fun isFiltered(packageName: String, notify: Notify, count: Boolean): Triplet<Boolean, Boolean, Boolean> {
        return isFiltered(packageName, notify.title, notify.text, count)
    }

    /**
     * @return do notify, do cancel, is filtered
     */
    fun isFiltered(
        packageName: String,
        title: String?,
        text: String?,
        count: Boolean
    ): Triplet<Boolean, Boolean, Boolean> {
        for (filter in filters) {
            if (filter.packageName == packageName) {
                return if (filter.isFiltered(title!!, text!!, count)) {
                    Triplet(false, filter.cancelFiltered, true)
                } else {
                    Triplet(filter.notifyUnfiltered, b = false, c = false)
                }
            }
        }
        return Triplet(false, b = false, c = false)
    }
}